### PR TITLE
Finalize refactoring of git repo

### PR DESCRIPTION
**Pull request by gtauzin**
Hi guys! So the PR is more or less ready, but :

I need your help with the following:
- [x] Check requirements versions (especially numba) requirements.txt + README.rst
- [x] Improve functions names (get is a getter => compute? st => steenrod?)
- [x] Check if steenroder structure is fine with you and that notebooks still run.
- [x] What do you think about renaming the package to pysteenroder? And transfering to Steenroder the cpp steenroder as steenroder (to follow naming in Ripyser/pyripser, flagser/pyflagser, etc...)

Please don't hesitate to push directly on the branch, I'll have less time in the coming week!

Optionally, to improve what is there, we could:
- [x] Put tutorial notebooks in docs (This has been surprisingly challenging even with nbsphinx...)
- [x] Put example notebooks in docs (maybe converting them first and storing them with git lfs?)
- [x] Improve rst files in docs
- [ ] Fix linting warnings (optional...)

Before, merging I need to:
- [x] Check doc action has been reverted to act on branch main

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+
+Installation
+------------
+
+Dependencies
+~~~~~~~~~~~~
+
+The latest stable version of ``steenroder`` requires:
+
+-  python (>= 3.8)
```
Comment (ulupo) [inline README.rst:36]: Can we not support 3.7?

**Review by ulupo — COMMENTED** [review]
Incredible work!  I'm afraid I won't have time for a full review this week, at least not until Thursday or Friday.  So, feel free to publish without my formal approval if a deadline needs to be met.

About versions:
Concerning `numba` versions -- 0.53.0 or later should suffice, that's what I was on when I started developing
Concerning `GUDHI` and `plotly` -- same versions as in `giotto-tda` should do
Concerning `psutil` -- no idea, but we only use `cpu_count` which is surely one of the oldest components in the library

File: README.rst (hunk 1 above)
Comment (gtauzin) [reply README.rst:36]: We can!

**Comment by ulupo** [comment]
> Improve functions names (get is a getter => compute? st => steenrod?)

I'm not attached to function names, feel free to liaise with @ammedmar.  

> Check if steenroder structure is fine with you and that notebooks still run.

Could you explain any backend changes you may have made to the functions previously in `__init__.py` and `plotting.py`?

> Improve rst files in docs

Can't do this myself in the short term, see above.

> What fo you think about renaming the package to pysteenroder? And transfering to Steenroder the cpp steenroder as steenroder (to follow naming in Ripyser/pyripser, flagser/pyflagser, etc...)

Fine with this!

**Comment by ulupo** [comment]
I also noticed that we don't always use `python -m pip install` but instead there are some occurrences of `pip3 install` and `pip install`.

**Comment by ulupo** [comment]
More importantly: I thought the recommended way to package nowadays is without `setup.py` files but using `.toml` files.

**Comment by ammedmar** [comment]
I refactored the notebooks I wrote to be compatible with the new structure. They all work but I am getting the following mysterious error from the cyclo-octane notebook:
`ValueError: numpy.ndarray size changed, may indicate binary incompatibility. Expected 88 from C header, got 80 from PyObject`
I pushed the changes to the notebooks to this branch as well as adding .idea/ to .gitignore.

I can take a crack at the names of functions and will report on this asap.

Regarding the name, I think the examples you mention are called `py(project_name)` because they are bindings for code originally written in `C++`. The situation seems different here since the jited `Python` code stands alone. I think that `pysteenroder` should be reserved for a binding of the code you wrote once it is tested and becomes the default. How do feel about calling it `cteenroder` in the meantime?

**Comment by gtauzin** [comment]
> Could you explain any backend changes you may have made to the functions previously in `__init__.py` and `plotting.py`?

No backend changes were made!

**Comment by gtauzin** [comment]
> More importantly: I thought the recommended way to package nowadays is without `setup.py` files but using `.toml` files.

At first, I was thinking of using poetry (with the pyproject.toml file), but that was too much to do to try it out. I am all for it, but I just did not have enough time to look into it. Happy to look into it after the first release.

**Comment by gtauzin** [comment]
> I also noticed that we don't always use `python -m pip install` but instead there are some occurrences of `pip3 install` and `pip install`.

That's true! All those files are automatically generated by github. I had not seen they were such a mess in the commands, I'll fix it.

**Comment by gtauzin** [comment]
> I refactored the notebooks I wrote to be compatible with the new structure. They all work but I am getting the following mysterious error from the cyclo-octane notebook: `ValueError: numpy.ndarray size changed, may indicate binary incompatibility. Expected 88 from C header, got 80 from PyObject` I pushed the changes to the notebooks to this branch as well as adding .idea/ to .gitignore.

No idea what is going on here, I did not change any backend code. Are you using the same dependencies?

> I can take a crack at the names of functions and will report on this asap.

That would be good, thanks! Please check that:
- __init__.py is updated
- notebooks are updated and run
- docs/api_reference.rst is updated
 
> Regarding the name, I think the examples you mention are called `py(project_name)` because they are bindings for code originally written in `C++`. The situation seems different here since the jited `Python` code stands alone. I think that `pysteenroder` should be reserved for a binding of the code you wrote once it is tested and becomes the default. How do feel about calling it `cteenroder` in the meantime?

I sent an email to discuss the choice. I am fine with any solution, but not the one that bet on changing packages names in the mid/long term. If we are to

File: README.rst (hunk 1 above)
Comment (gtauzin) [reply README.rst:36]: I'll make the change.

Comment (gtauzin) [reply README.rst:36]: Actually, we can't. The CI python 3.7 fails with:

```
 steenroder/core.py:44: in <module>
    @lru_cache
/opt/hostedtoolcache/Python/3.7.12/x64/lib/python3.7/functools.py:490: in lru_cache
    raise TypeError('Expected maxsize to be an integer or None')
E   TypeError: Expected maxsize to be an integer or None
```

**Comment by ammedmar** [comment]
I checked the notebooks and they run well. I also made the name changes as suggested by @gtauzin.

Regarding the incorporation of notebooks in the documentation. For [ComCh](https://comch.readthedocs.io/en/latest/index.html) I ended up using binder to accomplish something similar.

**Review by ammedmar — CHANGES_REQUESTED** [review]
If the notebooks are not incorporated into the docs in this PR, is there a role for clear_notebooks.py

**Review by ammedmar — COMMENTED** [review]
Thanks for the cleaning up. Some thoughts:
1. The user_guide.rst file is doing nothing and rasing a warning when building the docs. Is there a reason to keep it? 
2. Should the information in the README.rst be included in the docs?

**Review by ammedmar — APPROVED** [review]

**Comment by gtauzin** [comment]
> Thanks for the cleaning up. Some thoughts:
> 
>     1. The user_guide.rst file is doing nothing and rasing a warning when building the docs. Is there a reason to keep it?
> 
>     2. Should the information in the README.rst be included in the docs?

1. I left it there without including it to the docs as the next PR will fill it up with notebooks and it would have to be created again.
2. Yes, sure. I tried to do that but it was a bit complicated to have a proper index page structure. I can put it back if you like.

**Comment by ammedmar** [comment]
1. I think it would be best to add it in the next PR and not this one but I am not opposed to merging as is.
2. That would be good.
Thank you.

**Comment by gtauzin** [comment]
The notebooks are now in the docs and so is information of the readme (linking them in a good way is blocked because of security issues). You can have a look at the generated doc here: https://steenroder.github.io/steenroder/index.html

There are a few things that you should know:
- The notebooks are ran automatically and moved to the docs using nbmake and nbsphinx
- the klein bottle and cyclo-octane notebooks are not automatically ran by the CI. To be discplayed in the docs, they need to be locally ran and pushed to git. git LFS is used in the backend, so those big files won't be a problem. 
**NB: those two heavy notebooks are committed twice: once in notebooks/ with no output cells and once to docs/_notebooks/ran/ with all output cells.**

There are a few things that remain to do and I'll leave them to you:
- The rendering is not optimal because the headings levels in the notebooks are inconsistent (i.e. the `###` used for titles), so the table of content is broken. Notebooks need to be reworked to have a nice structure and consistent header levels.
- The cyclo-octane notebook is way too heavy to be ran. It's been running on my laptop for 2h and it's not even done doing VR. Please, if you can, run it fully and push the ran notebook in docs/_notebooks/ran/ without deleting the output cells.

**I suggest we merge this PR as is now, and I'd like to ask you to create a separate PR for the small remaining improvements.**

**Comment by ammedmar** [comment]
Hi @gtauzin 

Looking at the docs in the link you provided above shows the attached screenshot. Is that the behavior you expect?

Screenshot from 2022-04-18 17-40-48

**Comment by gtauzin** [comment]
Strange, it was working when I posted the link. The merge must have changed
something. I ll have a look.

**Pull request merged**